### ISSUE TITLE
A Simple model can not predict

### ISSUE THREAD

**Issue by jinfagang** [Dobiasd/frugally-deep#75]
I write a single keras model and test with some dummy data, the model trained and converted into fdeep model, but can not be inferenced on my Windows 10 platform.

```
import numpy as np
from keras.layers import Input, Dense, Conv2D, Flatten, MaxPool2D, AveragePooling2D, BatchNormalization, Activation, Convolution2D, Dropout, MaxPooling2D
from keras.models import Model, Sequential
import tensorflow as tf


nb_filters = 32
# size of pooling area for max pooling
pool_size = (2, 2)
# convolution kernel size
kernel_size = (3, 3)
nb_classes = 80
input_shape = (416, 416, 3)


def build_model():
    model = Sequential()
    model.add(Convolution2D(nb_filters, (kernel_size[0], kernel_size[1]),
                            padding='same',
                            input_shape=input_shape))  # 卷积层1
    model.add(Activation('relu'))  # 激活层
    model.add(Convolution2D(
        nb_filters, (kernel_size[0], kernel_size[1])))  # 卷积层2
    model.add(Activation('relu'))  # 激活层
    model.add(MaxPooling2D(pool_size=pool_size))  # 池化层
    model.add(Dropout(0.25))  # 神经元随机失活
    model.add(Flatten())  # 拉成一维数据
    model.add(Dense(128))  # 全连接层1
    model.add(Activation('relu'))  # 激活层
    model.add(Dropout(0.5))  # 随机失活
    model.add(Dense(nb_classes))  # 全连接层2
    model.add(Activation('softmax'))  # Softmax评分

    # 编译模型
    model.compile(loss='categorical_crossentropy',
                  optimizer='adadelta',
                  metrics=['accuracy'])
    return model


def main():
    model = build_model()

    with tf.Session() as sess:
        y = np.array([
            sess.run(tf.one_hot(2, 80)),
            sess.run(tf.one_hot(5, 80)),
            sess.run(tf.one_hot(2, 80)),
            sess.run(tf.one_hot(3, 80)),
            sess.run(tf.one_hot(2, 80)),
            sess.run(tf.one_hot(9, 80)),
        ])

    x = np.array(
        [
            np.random.random([416, 416, 3]),
            np.random.random([416, 416, 3]),
            np.random.random([416, 416, 3]),
            np.random.random([416, 416, 3]),
            np.random.random([416, 416, 3]),
            np.random.random([416, 416, 3]),
        ]
    )
    print(x)
    print(y)

    model.fit(
        x=x,
        y=y,
        epochs=30
    )

    model.summary()

    model.save('keras_model.h5', include_optimizer=False)


if __name__ == '__main__':
    main()

```

The error says something like 
```
Debug Error!

abort() has been called..
```
this happends right in the load_model line,

**Comment by Dobiasd** [comment]
I just did the following steps:

Saved [your code](https://gist.github.com/Dobiasd/16edaf183609bef3864a8235f80583b0) as `save_model.py`

Ran it to generate `keras_model.h5`:
```bash
python3 save_model.py
```

Converted it to `fdeep_model.json`:
```bash
python3 ./keras_export/convert_model.py keras_model.h5 fdeep_model.json
```

Compiled the following `main.cpp`:
```cpp
#include <fdeep/fdeep.hpp>
int main()
{
    const auto model = fdeep::load_model("fdeep_model.json");
    std::cout << "Done." << std::endl;
}

```

like this:
```bash
g++ -std=c++14 -O3 main.cpp -o main
```

And ran the resulting executable:
```bash
./main
```

And indeed, it crashes:
```
Loading json ... done. elapsed time: 12.562739 s
Running test 1 of 1 ... done. elapsed time: 0.334693 s
terminate called after throwing an instance of 'std::runtime_error'
  what():  test failed: output=0 pos=2,0,0 value=-nan target=1
Aborted
```

It works however if I add `#define FDEEP_FLOAT_TYPE double`, like so:
```cpp
#define FDEEP_FLOAT_TYPE double
#include <fdeep/fdeep.hpp>
int main()
{
    const auto model = fdeep::load_model("fdeep_model.json");
    std::cout << "Done." << std::endl;
}
```

Could you please try if this helps on your system too?
If so, we can discuss the reasons behind this afterwards. :slightly_smiling_face:

**Comment by jinfagang** [comment]
@Dobiasd Thanks!!!!! Seems worked, but what does this float do and why?

**Comment by Dobiasd** [comment]
When `convert_model.py` writes `fdeep_model.json` it runs a forward pass with random input data, and stores this data (input and output) with the model.

In C++ `fdeep::load_model` not only reads the model architecture and the weights but also this test data (input and output). It then does a prediction with the example input and compares the C++ output with the output that was generated in Python. That way frugally-deep guarantees that your model behaves exactly the same in C++ as it does in Python.

Your model however was trained with random data too. That way the softmax layer was unable to learn anything meaningful and we run into a `nan` situation (perhaps something like division by almost zero).

When you actually train your model normally, this should not be a problem any more. You can also skip this verification process by setting the `verify` parameter to `false` in your call to `fdeep::load_model`.

Setting `#define FDEEP_FLOAT_TYPE double` makes fdeep use `double` instead of `float`. This also makes the problem go away since the added precision results in the calculation to end up with a non-`nan` value. But usually this should not be needed. It is just because of the bogus weights stored with the randomly trained model.

So in general any of the following three things should help (in order of reasonability)
1) Actually train the model
2) Skip the verification process when loading the model in C++
3) Set `FDEEP_FLOAT_TYPE` to `double`

So solution 3 now worked in your case, but I'll invite you to try out the two (more sane) solutions too. :wink:

**Comment by jinfagang** [comment]
@Dobiasd Many helpful indeed. I'll test those situations.

**Comment by chammika-become** [comment]
@jinfagang the error is caused by `float` value overflow in softmax layer summation.

https://github.com/Dobiasd/frugally-deep/blob/master/include/fdeep/layers/softmax_layer.hpp#L44

`sum ==> inf` makes subsequent division  `v ==> nan`.

@Dobiasd  The assertion in line `46` can be updated to `assertion(sum != 0 && !std::isinf(sum) ` to catch this and make the helpful suggestion to change the type to `double` as you mentioned above.
Is there a reason not to have the `double` as the default type ?

**Comment by Dobiasd** [comment]
@chammika-become Wow, thanks for this good analysis. Would you like to make the fix as a PR? Otherwise I of course can do it too.

The main reason for `float` as default type is simply the memory consumption of a loaded model.

**Comment by chammika-become** [comment]
Ok, I get the point about `float` . I stop short of making a PR because I thought this might add an extra overhead at the summation step (note that the && won't short the expression evaluation in assert when most of the time sum!=0 in regular computation).

**Comment by chammika-become** [comment]
@Dobiasd  Made a PR anyway if you think it's OK please merge
 https://github.com/Dobiasd/frugally-deep/pull/77
